### PR TITLE
Close Channels at DFK cleanup

### PR DESCRIPTION
**Pull request by benclifford**
Following on from #3403, which tightens the type signature of unused Channel.close(), this PR makes the DFK actually call Channel.close() when it shuts down an executor.

At startup, the DFK chooses whether to manage things with channels by checking for the presence of a `script_dir` attribute on the relevant provider (!) and this PR sticks with that same informal protocol.

On the LocalChannel, this has no effect. On the SSH Channels, this shuts down the SSH connection and removes an otherwise-abandoned thread.

# Changed Behaviour

Channels will be closed at shutdown.

## Type of change

- Bug fix

### CODE REVIEW

**Comment by benclifford** [comment]
this is consistently(?) hanging in the flux tests... (in a similar way to the very-often-hanging flux tests, but seemly more so). I'll try to debug.

**Comment by benclifford** [comment]
(crossreference #3484 for main issue to track flux CI hangs)

File: parsl/dataflow/dflow.py [hunk 1]
```diff
@@ -1277,6 +1277,19 @@ def cleanup(self) -> None:
...
+            if hasattr(executor, 'provider'):
+                if hasattr(executor.provider, 'script_dir'):
+                    logger.info(f"Closing channel(s) for {executor.label}")
+
+                    if hasattr(executor.provider, 'channels'):
+                        for channel in executor.provider.channels:
+                            channel.close()
+                    else:
+                        assert hasattr(executor.provider, 'channel'), "If provider has no .channels, it must have .channel"
+                        executor.provider.channel.close()
```
Comment (khk-globus) [inline parsl/dataflow/dflow.py:1289]: For debugging / retrospective, would `log.debug()` for each closed channel be of any use?  (And if so, do we have recognizable output from `channel.__repr__` or `.__str__` to make that feasible?)

**Review by khk-globus — APPROVED** [review]

File: parsl/dataflow/dflow.py (hunk 1 above)
Comment (benclifford) [reply parsl/dataflow/dflow.py:1289]: yes to both: logging around closing is pretty useful here, even at INFO level, due to Parsl's history of hanging at shutdown; Channels are rendered as a recursive part of the start log-the-`Config`-object so they "should" already have an appropriate repr().

**Pull request merged**